### PR TITLE
JAMES PHILLIPS MBP AUTH-I

### PR DESCRIPTION
**Pull request by CASDFJ**
# Authentication Project

## Topics

- Authentication.
- Express Middleware.
- Password Hashing.

## Assignment

**Part one, due Monday**: Use Node.js, Express and Knex to build an API that provides **Register** and **Login** functionality using SQLite to store _User_ information. Make sure the password is not stored as plain text.

**Part two, due Tuesday**: Use **sessions** and **cookies** to keep a record of logged in users across requests.

## Instructions

You will build the solution from scratch, no starter code is provided. Feel free to structure your API anyway you want, but aim at making it easy to maintain in the future.

### Download Project Files

- **Fork** and **Clone** this repository.
- **CD into the folder** where you cloned the repository.
- Do your magic!

### Implement Requirements

- Take the steps necessary to create a `package.json` to keep a record of all dependencies.
- Configure an _npm script_ named _"start"_ that will execute your code using _nodemon_ so that the server restarts on changes. Make _nodemon_ be a development time dependency only, it shouldn't be deployed to production.
- Design and build a set of endpoints listed below.
- **Use _Postman_ to test the API as you work through the exercises.**

#### Endpoints

| Method | Endpoint      | Description                                                                                                                                                                                                                                                                                 |
| ------ | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| POST   | /api/register | Creates a `user` using the information sent inside the `body` of the request. **Hash the password** before saving the user to the database.                                                                                                                                                 |
| POST   | /api/login    | Use the credentials sent inside the `body` to authenticate the user. On successful login, create a new session for the user and send back a 'Logged in' message and a cookie that contains the user id. If login fails, respond with the correct status code and the message: 'You shall not pass!' |
| GET    | /api/users    | If the user is logged in, respond with an array of all the users contained in the database. If the user is not logged in repond with the correct status code and the message: 'You shall not pass!'. Use this endpoint to verify that the password is hashed before it is saved.            |

## Stretch Problem

- Write a piece of **global** middleware that ensures a user is logged in when accessing _any_ route prefixed by `/api/restricted/`. For instance, `/api/restricted/something`, `/api/restricted/other`, and `/api/restricted/a` should all be protected by the middleware; only logged in users should be able to access these routes.
- Build a React application that implements components to register, login and view a list of users. Gotta keep sharpening your React skills.

### CODE REVIEW

File: database/dbConfig.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+const knex = require('knex');
+
+const knexConfig = require('../knexfile.js');
+
+module.exports = knex(knexConfig.development);
```
Comment (Mister-Corn) [inline database/dbConfig.js]: This is kinda slick. What inspired you to do this? 😲

File: index.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+server.post("/api/login", (req, res) => {
+  const creds = req.body;
+
+  db("users")
+    .where({ username: creds.username })
+    .first()
+    .then(user => {
+      if (user && bcrypt.compareSync(creds.password, user.password)) {
+        // grab roles from user
+        // req.session.roles = roles
```
Comment (Mister-Corn) [inline index.js]: Good comment for authorization

File: index.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        res.status(401).json({ Error: "Cannot Authorize" });
+      }
+    })
+    .catch(err => {
+      console.log("Error: ", err);
+      res.status(500).json({ Error: "Login Failed" });
+    });
+});
+//============POST LOGIN ENDPOINT============//
+
```
Comment (Mister-Corn) [inline index.js]: Great work so far. 👍 I'll put in a PR comment once we're in the next project.

File: index.js [hunk 4]
```diff
@@ -1,17 +1,38 @@
...
+const server = express();
+
+const login = require("./Login/login.js");
+const configMiddleware = require("./config/middleware.js");
+
+//configure middleware for the server
+configMiddleware(server);
+
+server.get("/download", (req, res, next) => {
```
Comment (Mister-Corn) [inline index.js]: Still not exactly sure what this endpoint is doing.

File: index.js [hunk 5]
```diff
@@ -35,92 +56,29 @@ const sessionsConfig = {
 //======session configuration====//
 
 server.use(session(sessionsConfig));
-server.use(express.json());
-server.use(cors());
```
Comment (Mister-Corn) [inline index.js]: Hopefully you're getting a better hang of this middleware queue 👍

File: index.js [hunk 6]
```diff
@@ -35,92 +56,29 @@ const sessionsConfig = {
...
+server.use("/api", login);
 
 //========middleware==========//
-function auth(req, res, next) {
-  if (req.session && req.session.username) {
-    next();
-  } else {
-    res.status(401).json({ Error: "You shall not pass!!" });
+function errorHandler(err, req, res, next) {
+  console.error(err);
```
Comment (Mister-Corn) [inline index.js]: This `errorHandler` function is new to me, so I'm legitimately curious. How does it work? When does express call it?

**Comment by Mister-Corn** [comment]
Good work. 👍 Hopefully you're getting a better grasp on the middleware queue. I would like to know how that `errorHandler` middleware works. That looks pretty cool!

File: index.js (hunk 6 above)
Comment (CASDFJ) [reply index.js]: The errorHandler is just middle ware  for when the get(/download) doesn't work

File: index.js (hunk 4 above)
Comment (CASDFJ) [reply index.js]: The /download is used to connect the routers with the configuration files and index.js. It was also something referenced by Luis in the videos a while back. Same with the error handler.

**Pull request closed without merging**